pageserver/src/storage_sync/download.rs

```diff
@@ -429,6 +365,121 @@ where
     }
 }
 
+async fn download_storage_object(
+    storage: &GenericRemoteStorage,
+    to_path: &Path,
+) -> Result<Download, DownloadError> {
+    async fn do_download_storage_object<P, S>(
+        storage: &S,
+        to_path: &Path,
+    ) -> Result<Download, DownloadError>
+    where
+        P: std::fmt::Debug + Send + Sync + 'static,
+        S: RemoteStorage<RemoteObjectId = P> + Send + Sync + 'static,
+    {
+        let remote_object_path = storage
+            .remote_object_id(to_path)
+            .with_context(|| {
+                format!(
+                    "Failed to get the storage path for target local path '{}'",
+                    to_path.display()
+                )
+            })
+            .map_err(DownloadError::BadInput)?;
+
+        storage.download(&remote_object_path).await
+    }
+
+    match storage {
+        GenericRemoteStorage::Local(storage) => do_download_storage_object(storage, to_path).await,
+        GenericRemoteStorage::S3(storage) => do_download_storage_object(storage, to_path).await,
+    }
+}
+
+async fn get_timeline_sync_ids(
```

Comment:
I've found a good use case, where such "associated type polymorphism" could simplify our code generally: sync id resolution was the only place where `RemoteObjectName` trait was used.

Yet, instead of introducing an extra trait entity, we could extract the object name in place.